Test/Tests/RealtimeClientConnectionTests.swift

```diff
@@ -3089,6 +3180,372 @@ class RealtimeClientConnectionTests: XCTestCase {
+            client.connection.once(.connected) { stateChange in
+                expect(client.connection.createRecoveryKey()).notTo(beNil())
+                client.internal.onError(ARTProtocolMessage())
+            }
+            client.connection.once(.failed) { stateChange in
+                expect(client.connection.createRecoveryKey()).to(beNil())
+                done()
+            }
+        }
+    }
+
+    // RTN16f
+    func test__085__Connection__Connection_recovery__should_set_internal_message_serial_to_component_in_recovery_key() throws {
+        let test = Test()
+        let options = try AblyTests.commonAppSetup(for: test)
+        let client = AblyTests.newRealtime(options).client
+        defer { client.dispose(); client.close() }
+
+        waitUntil(timeout: testTimeout) { done in
+            client.connection.once(.connected) { stateChange in
+                done()
+            }
+        }
+        expect(client.connection.createRecoveryKey()).toNot(beNil())
+        
+        let recoverOptions = options
+        recoverOptions.recover = client.connection.createRecoveryKey()
+        let recoveryKey = ARTConnectionRecoveryKey.fromJsonString(client.connection.createRecoveryKey()!)
+        let recoverClient = AblyTests.newRealtime(recoverOptions).client
+        expect(recoverClient.internal.msgSerial).to(equal(recoveryKey.msgSerial))
+    }
+    
+    // RTN16j
+    func test__086__Connection__Connection_recovery__library_should_create_channel_with_corresponding_serial_in_given_recovery_key() throws {
+        let test = Test()
+        let options = try AblyTests.commonAppSetup(for: test)
+
+        let client = AblyTests.newRealtime(options).client
+        defer { client.dispose(); client.close() }
+        
+        let channelName = test.uniqueChannelName()
+        var expectedChannelSerial: String?
+        waitUntil(timeout: testTimeout) { done in
+            let partialDone = AblyTests.splitDone(2, done: done)
+            client.connection.once(.connected) { stateChange in
+                let channel = client.channels.get(channelName)
+                channel.on(.attached) { _ in
+                    expectedChannelSerial = channel.internal.serial
+                    partialDone()
+                }
+                channel.attach()
+                partialDone()
+            }
+        }
+
+        let recoverOptions = options
+        recoverOptions.recover = client.connection.createRecoveryKey()
+        
+        let recoveredClient = AblyTests.newRealtime(recoverOptions).client
+        defer { recoveredClient.dispose(); recoveredClient.close() }
+        expect(recoveredClient.connection.state).toEventually(equal(.connected), timeout: testTimeout)
+        
+        let recoveredChannel = recoveredClient.channels.get(channelName)
+        expect(recoveredChannel).toNot(beNil())
+        expect(recoveredChannel.internal.serial).to(equal(expectedChannelSerial))
+    }
+  
+    // RTN16k
```

Comment:
I couldn't see a test for the "Once the library has successfully connected to Ably, it should never again supply a `recover` querystring param" part of RTN16k.

Reply:
7e45869f06b1faaf78a0d48d3f247f23b5d16875